### ISSUE TITLE
Consider ignoring common Tooling tables eg __RefactorLog when doing reveng

### ISSUE THREAD

**Issue by mikes-gh** [aspnet/EntityFramework#5126]
2 examples I have in my reveng generated code.
__RefactorLog (used my SSDT)
dtproperties (not sure about this one I think tool is now dead so maybe moot).

Consider an ignore list for reveng 
Not sure how __EFMigrationHistory is ignored.
I cant see any scenario where the above would be useful in the dbcontext and model classes.

**Comment by natemcmaster** [comment]
As you may know, dbcontext scaffolding can [take parameters](http://docs.efproject.net/en/latest/cli/powershell.html#scaffold-dbcontext) to filter on specific tables and schema. For now I consider this the best way to prevent these tables from being scaffolded. To solve this generally, we'd need a clear answer to "what are common table names that should always be excluded?"

**Comment by ErikEJ** [comment]
Providing exclusion filter rather than inclusion filter would be a good start

**Comment by rowanmiller** [comment]
We should handle this with the ability to exclude tables

**Comment by rowanmiller** [comment]
BTW we should be ignoring system tables... not sure if __RefactorLog is a system table or not

**Comment by ErikEJ** [comment]
You should add: is_ms_shipped to the tables query - and refactorlog is a user table: 

https://technet.microsoft.com/da-dk/library/ms190324(v=sql.110).aspx 

http://stackoverflow.com/questions/10029979/how-to-check-if-sql-server-tables-are-system-tables

**Comment by mikes-gh** [comment]
@ErikEJ 
Good find
I confirm both the examples I quoted are in system table folder 

![system tables](https://cloud.githubusercontent.com/assets/16208742/14691442/2a30ac34-0749-11e6-98c8-fc410a740be8.png)

so the stack overflow article is the answer.  Although it doesn't look as straightforward as is_ms_shipped = 0 

> it's a combination - either is_ms_shipped from sys.tables being 1, or having a particular extended property set.

The article shows the query SSMS uses to classify.

**Comment by mikes-gh** [comment]
**The following**
```
select Name
from sys.tables AS tbl
where 
	CAST(
	case 
		when tbl.is_ms_shipped = 1 then 1
		when (
			select major_id 
			from sys.extended_properties 
			where 
				major_id = tbl.object_id and 
				minor_id = 0 and 
				class = 1 and 
				name = N'microsoft_database_tools_support') 
			is not null then 1
		else 0
	end AS Bit) = 0   
```
Gives a list of user tables excluding the pseudo System Tables

This is based on what SSMS spits out and the stack overflow article.
I tried to make it pretty honestly..

**Comment by mikes-gh** [comment]
Could be an idea to make __EFMigrationHistory  is_ms_shipped = 1 aswell?

http://www.tech-recipes.com/rx/49940/is_ms_shipped-property-in-sql-server/

Probably SSDT should be using it aswell for __RefactorLog

**Comment by ErikEJ** [comment]
I am not sure a user can set is_ms_shipped = 1

**Comment by mikes-gh** [comment]
@ErikEJ  I was thinking the EF team (rather than an end user) could set __EFMIgrationHistory to is_ms_shipped = 1.
Presumably you could speak to the SQL Server team to see how to do this or if its a good idea.
Doesn't solve the __RefactorLog case though which AFAIK can only be determined from the horrible query above.

**Comment by ErikEJ** [comment]
Setting is_ms_shipped would require EF (on behalf of the dev/person running migrations) to use an undocumented stored procedure - I would strongly advise against doing that! http://sqlblog.com/blogs/kalen_delaney/archive/2008/08/10/geek-city-system-objects.aspx

**Comment by mikes-gh** [comment]
Fair point forgot it runs in user context.  Its also filtered by default anyway.

`ToolingTableWhereClause` here?  Similar to TemporalTableWhereClause

https://github.com/aspnet/EntityFramework/blob/2acd2f1f250285f2fe9bb9c81c27f7bb11355e6b/src/Microsoft.EntityFrameworkCore.SqlServer.Design/SqlServerDatabaseModelFactory.cs#L200

**Comment by ErikEJ** [comment]
I think it should be done on the column query level..

And a PR should include both the tooling condition and the is_ms_shipped condition

**Comment by mikes-gh** [comment]
That would be simplest..  My vote is for where clause and keep --Tables as is.
I think you are more likely to want to include rather than exclude.

I'm not adverse to the exclude option but it would involve breaking things to do it properly.

- Remove `--Tables` Add `--IncludeTables` and `--ExcludeTables` 
- You would still have to additionally exclude the tooling tables by default anyway as having to remeber to type 
`--Exclude "dtproperties,__RefactorLog"` would be a pain.
- Handle priority of include vs exclude in case someone puts a table in both.

**Comment by lajones** [comment]
Discussed offline and decided to go with just updating the query as suggested by @mikes-gh above (thanks for that) and leaving the flags as is for now. I.e. if we see enough need we'll add the `ExcludeXYZ` command line options later.

**Comment by lajones** [comment]
Fixed as discussed above with PR #5213.
